src/isql/isql.epp

```diff
@@ -378,34 +383,42 @@ static int win32ReadConsole(FILE* file, char* buffer, size_t bufferSize)
 				break;
 		}
 
-		WCHAR wideBuf[2];
-		DWORD charsRead;
+		if (!charsRead)
+		{
+			if (!ReadConsoleW(handle, wideBuf, MAX_LINE_LENGTH, &charsRead, NULL))
+				return -1;
 
-		// Reading one code unit at a time is inefficient, but since this code
-		// is used only for the interactive console, that shouldn't matter.
-		if (!ReadConsoleW(handle, wideBuf, 1, &charsRead, 0))
-			return -1;
+			if (!charsRead)
+				break;
 
-		if (!charsRead)
-			break;
+			pWide = wideBuf;
+		}
 
 		DWORD wideBufLen = 1;
+		charsRead--;
 
-		if (wideBuf[0] >= 0xD800 && wideBuf[0] <= 0xDBFF)
+		if (IS_HIGH_SURROGATE(*pWide))
 		{
 			// High surrogate, read one more code unit.
-			if (!ReadConsoleW(handle, wideBuf + 1, 1, &charsRead, 0))
-				return -1;
-
 			if (charsRead)
-				++wideBufLen;
+				charsRead--;
+			else
+			{
+				DWORD read2;
+				if (!ReadConsoleW(handle, pWide + 1, 1, &read2, NULL))
```

Comment:
You already read whole line into `wideBuf`, there is nothing to read here. `WideCharToMultiByte` below will take care about surrogates automatically. Whole this `if` is meaningless AFAIK (as well as outer loop, perhaps - you just read the line and convert it, no piecewise read is possible).

Reply:
The code is generic and doesn't assume that whole line is read by single call of `ReadConsoleW()`. 
I see no need to break this.

Reply:
First, surrogates aren't supported in modern Unicode standard at all.

Second, you cannot enter them from console in a way that only a half of surrogate pair is been read exactly because of the issue: the console disallow to enter more characters than fits the requested buffer. For old code such situation was theoretically possible because it read by tiny pieces and the system console buffer was bigger than requested buffer. In the new code this call will wait for the next line requesting one character which, in turn, will again limit such continuation line by 254 characters, no?.. If you want to preserve such logic, move the incomplete character into the beginning of the buffer and read `MAX_LINE_LENGTH - 1` here at least.

Reply:
I've tested the code with small buffer and actually able to enter surrogate pair in the way you consider impossible.
There is no documenattion and no guarantee that next Windows version keeps internal buffer of 255 characters, so I prefer to keep existing working already tested code.
I see no reason to convince you here, it is senseless.

PS ask yourself - why wideBuf is declared with MAX_LINE_LENGTH +1 length

Reply:
> First, surrogates aren't supported in modern Unicode standard at all.

That is not correct, Unicode 16 (the latest version) still covers surrogates, and explicitly says they should only be used with UTF-16, and UTF-16 needs surrogates to be able to encode all characters defined by Unicode. And guess what `ReadConsoleW` produces: UTF-16.

Reply:
And I'd guess that pasting a U+10000 or higher character in a console that supports it will produce surrogate pairs.